core/src/main/java/com/yahoo/imapnio/client/ImapClientRespReader.java

```diff
@@ -1,122 +1,148 @@
-                        }
-                    }
+    protected Object decode(final ChannelHandlerContext ctx, final ByteBuf inputBuf) throws Exception {
+
+        while (inputBuf.readableBytes() > 0) {
+
+            if (literalCount <= 0) { // LINE mode - read until CRLF
+
+                final ByteBuf lineBuf = (ByteBuf) super.decode(ctx, inputBuf); // Read a CRLF terminated line from inputBuf
+                if (lineBuf == null) { // no CRLF seen in this case, don't return existing buffer since it is not done
+                    return null;
+                }
+                // Now lets check for literals : {<digits>}CRLF
+                // Note: index needs to >= 5 for the above sequence to occur
+                final int lineLen = lineBuf.readableBytes();
+                final boolean noLiterals = (lineLen < FIVE || lineBuf.getByte(lineLen - THREE) != '}'); // getByte() is peek operation
+
+                if (noLiterals) { // enough for a complete IMAP response, return the data
+                    return getFinalResponse(lineBuf);
                 }
-            } while (literalCount > 0 && buffer.readableBytes() > 0 && lineBuffer != null);
-
-            if (literalBuffer != null && literalCount == 0) {
-                // Return literal mode result.
-                final ByteBuf lastLineBuffer = (ByteBuf) super.decode(ctx, buffer);
-                final int lastLineIndex = lastLineBuffer.readableBytes();
-                final byte[] lastLineBytes = new byte[lastLineIndex];
-                lastLineBuffer.getBytes(lastLineBuffer.readerIndex(), lastLineBytes);
-                literalBuffer.writeBytes(lastLineBytes);
-                final ByteBuf result = literalBuffer.copy();
-                literalBuffer = null;
-                return result;
-            } else {
-                // Return line mode result.
-                return lineBuffer;
+
+                // extracting literal count between curly braces
+                this.literalCount = getLiteralCount(lineBuf, lineLen);
+                if (this.literalCount < 0) { // Nope, not a literal ?
+                    return getFinalResponse(lineBuf);
+                }
+
+                // literals follows if reaching here
+                if (literalBuf == null) {
+                    literalBuf = Unpooled.buffer(lineLen + literalCount + EXTRA_PADDING_LEN);
+                }
+                literalBuf.writeBytes(lineBuf); // add current line (ex: "* 1 FETCH (FLAGS (\Seen $NotJunk) BODY[] {4495}\r\n")
+                // back to top of loop to enter literal mode block
+
+            } else { // LITERAL mode - read till reaching the count or end of inputBuf
+                final int avail = inputBuf.readableBytes(); // available bytes unread in inputBuf
+                final int actual = (literalCount <= avail) ? literalCount : avail; // actual length to copy
+                literalBuf.writeBytes(inputBuf, actual);
```

Comment:
when client does not send the data, DefaulteChannelPipeline.fireChannelRead() will not be called.  Following is the stack when client sends the data.

Thread [nioEventLoopGroup-2-1] (Suspended (breakpoint at line 75 in ImapClientRespReader))	
	ImapClientRespReader.decode(ChannelHandlerContext, ByteBuf) line: 75	
	ImapClientRespReader(DelimiterBasedFrameDecoder).decode(ChannelHandlerContext, ByteBuf, List<Object>) line: 216	
	ImapClientRespReader(ByteToMessageDecoder).callDecode(ChannelHandlerContext, ByteBuf, List<Object>) line: 411	
	ImapClientRespReader(ByteToMessageDecoder).channelRead(ChannelHandlerContext, Object) line: 248	
	DefaultChannelHandlerContext(AbstractChannelHandlerContext).invokeChannelRead(Object) line: 372	
	AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext, Object) line: 358	
	DefaultChannelHandlerContext(AbstractChannelHandlerContext).fireChannelRead(Object) line: 350	
	IdleStateHandler.channelRead(ChannelHandlerContext, Object) line: 266	
	DefaultChannelHandlerContext(AbstractChannelHandlerContext).invokeChannelRead(Object) line: 372	
	AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext, Object) line: 358	
	DefaultChannelHandlerContext(AbstractChannelHandlerContext).fireChannelRead(Object) line: 350	
	SslHandler.unwrap(ChannelHandlerContext, ByteBuf, int, int) line: 1066	
	SslHandler.decode(ChannelHandlerContext, ByteBuf, List<Object>) line: 900	
	SslHandler(ByteToMessageDecoder).callDecode(ChannelHandlerContext, ByteBuf, List<Object>) line: 411	
	SslHandler(ByteToMessageDecoder).channelRead(ChannelHandlerContext, Object) line: 248	
	DefaultChannelHandlerContext(AbstractChannelHandlerContext).invokeChannelRead(Object) line: 372	
	AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext, Object) line: 358	
	DefaultChannelPipeline$HeadContext(AbstractChannelHandlerContext).fireChannelRead(Object) line: 350	
	DefaultChannelPipeline$HeadContext.channelRead(ChannelHandlerContext, Object) line: 1334	
	DefaultChannelPipeline$HeadContext(AbstractChannelHandlerContext).invokeChannelRead(Object) line: 372	
	AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext, Object) line: 358	
	DefaultChannelPipeline.fireChannelRead(Object) line: 926	
	NioSocketChannel$NioSocketChannelUnsafe(AbstractNioByteChannel$NioByteUnsafe).read() line: 129	
	NioEventLoop.processSelectedKey(SelectionKey, AbstractNioChannel) line: 610	
	NioEventLoop.processSelectedKeysOptimized(SelectionKey[]) line: 551	
	NioEventLoop.processSelectedKeys() line: 465	
	NioEventLoop.run() line: 437	
	SingleThreadEventExecutor$5.run() line: 873	
	DefaultThreadFactory$DefaultRunnableDecorator.run() line: 144	
	FastThreadLocalThread(Thread).run() line: 748	
Daemon Thread [threadDeathWatcher-3-1] (Running)	

===========================================
In ByteToMessageDecoder, it will continue call decode till the input ByteBuf isReadable is false (aka reading of the buf reaches the end)

Reply:
discussed/showed with UT on how ByteToMessageDecoder.callDecode() behaves in terms of handling the input ByteBuf.  In a nutshell, it will break out of the loop if  handler does not consume any input and not generate any output.  It will not hold up the thread in the endless loop.  line 421 of ByteToMessageDecoder:
               if (outSize == out.size()) {
                    if (oldInputLength == in.readableBytes()) {
                        break;
                    } else {
                        continue;
                    }
                }